File: src/context.js
import { randomUUID } from "node:crypto";
import { DEFAULT_CONTEXT_TTL_MS } from "./constants.js";
import { applyContextByteLimit } from "./limits.js";
import { validateContextMode } from "./modes.js";
import { hashContent } from "./hash.js";
import { classifyContextSource, normalizeProvenance } from "./provenance.js";
import { applyContextRedaction } from "./redaction.js";
import { assertNonEmptyString, assertPlainObject } from "./validation.js";

export function normalizeContext(input, options = {}) {
  assertPlainObject(input, "input");
  const now = options.now ?? new Date();
  const capturedAt = input.capturedAt ?? now.toISOString();
  const expiresAt =
    input.expiresAt ?? new Date(now.getTime() + (options.ttlMs ?? DEFAULT_CONTEXT_TTL_MS)).toISOString();

  validateContextMode(input.contextMode);
  assertNonEmptyString(input.tenantId, "input.tenantId");
  assertNonEmptyString(input.userId, "input.userId");
  assertNonEmptyString(input.sessionId, "input.sessionId");
  assertNonEmptyString(input.provider, "input.provider");
  assertPlainObject(input.resourceRef, "input.resourceRef");
  assertNonEmptyString(input.resourceRef.resourceId, "input.resourceRef.resourceId");
  assertNonEmptyString(input.content, "input.content");

  const redactionResult = applyContextRedaction(input.content, {
    policy: options.redactionPolicy,
    rules: options.redactionRules
  });
  const limitResult = applyContextByteLimit(redactionResult.content, {
    maxBytes: options.maxBytes,
    oversizedBehavior: options.oversizedBehavior ?? "truncate",
    rejectionReason: "context content exceeds configured safe limit"
  });
  const classification = classifyContextSource(input);
  const provenance = normalizeProvenance({
    ...input,
    capturedAt,
    resourceId: input.resourceRef.resourceId,
    resourceVersion: input.resourceRevision ?? input.resourceVersion ?? null,
    selectionAnchor: input.anchors?.selectionAnchor ?? null
  });

  return {
    contextId: input.contextId ?? randomUUID(),
    tenantId: input.tenantId,
    userId: input.userId,
    sessionId: input.sessionId,
    provider: input.provider,
    resourceRef: input.resourceRef,
    contextMode: input.contextMode,
    sourceType: classification.sourceType,
    trustLevel: classification.trustLevel,
    content: limitResult.content,
    contentHash: hashContent(limitResult.content),
    anchors: input.anchors ?? {},
    resourceRevision: input.resourceRevision ?? null,
    metadata: {
      ...(input.metadata ?? {}),
      redaction: redactionResult.metadata,
      contentLimit: limitResult.metadata
    },
    provenance,
    capturedAt,
    expiresAt,
    clientSupplied: classification.clientSupplied,
    connectorVerified: classification.connectorVerified
  };
}

File: src/ai_assist_context_service/context.py
from datetime import datetime, timedelta, timezone
from uuid import uuid4

from .constants import DEFAULT_CONTEXT_TTL_MS
from .hash import hash_content
from .limits import apply_context_byte_limit
from .modes import validate_context_mode
from .provenance import classify_context_source, normalize_provenance
from .redaction import apply_context_redaction
from .validation import assert_non_empty_string, assert_plain_object


def normalize_context(input_value, options=None):
    options = options or {}
    assert_plain_object(input_value, "input")
    now = options.get("now", datetime.now(timezone.utc))
    captured_at = input_value.get("capturedAt", _to_iso(now))
    ttl_ms = options.get("ttlMs", DEFAULT_CONTEXT_TTL_MS)
    expires_at = input_value.get("expiresAt", _to_iso(now + timedelta(milliseconds=ttl_ms)))

    validate_context_mode(input_value.get("contextMode"))
    assert_non_empty_string(input_value.get("tenantId"), "input.tenantId")
    assert_non_empty_string(input_value.get("userId"), "input.userId")
    assert_non_empty_string(input_value.get("sessionId"), "input.sessionId")
    assert_non_empty_string(input_value.get("provider"), "input.provider")
    assert_plain_object(input_value.get("resourceRef"), "input.resourceRef")
    assert_non_empty_string(input_value["resourceRef"].get("resourceId"), "input.resourceRef.resourceId")
    assert_non_empty_string(input_value.get("content"), "input.content")

    redaction_result = apply_context_redaction(
        input_value["content"],
        {
            "policy": options.get("redactionPolicy"),
            "rules": options.get("redactionRules"),
        },
    )
    limit_result = apply_context_byte_limit(
        redaction_result["content"],
        {
            "max_bytes": options.get("maxBytes"),
            "oversized_behavior": options.get("oversizedBehavior", "truncate"),
            "rejection_reason": "context content exceeds configured safe limit",
        },
    )
    classification = classify_context_source(input_value)
    anchors = input_value.get("anchors") or {}
    provenance = normalize_provenance(
        {
            **input_value,
            "capturedAt": captured_at,
            "resourceId": input_value["resourceRef"]["resourceId"],
            "resourceVersion": input_value.get("resourceRevision") or input_value.get("resourceVersion"),
            "selectionAnchor": anchors.get("selectionAnchor"),
        }
    )

    return {
        "contextId": input_value.get("contextId") or str(uuid4()),
        "tenantId": input_value["tenantId"],
        "userId": input_value["userId"],
        "sessionId": input_value["sessionId"],
        "provider": input_value["provider"],
        "resourceRef": input_value["resourceRef"],
        "contextMode": input_value["contextMode"],
        "sourceType": classification["sourceType"],
        "trustLevel": classification["trustLevel"],
        "content": limit_result["content"],
        "contentHash": hash_content(limit_result["content"]),
        "anchors": anchors,
        "resourceRevision": input_value.get("resourceRevision"),
        "metadata": {
            **(input_value.get("metadata") or {}),
            "redaction": redaction_result["metadata"],
            "contentLimit": limit_result["metadata"],
        },
        "provenance": provenance,
        "capturedAt": captured_at,
        "expiresAt": expires_at,
        "clientSupplied": classification["clientSupplied"],
        "connectorVerified": classification["connectorVerified"],
    }


def _to_iso(value):
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
